File: 131-palindrome-partitioning/131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        vector<string> path;
        func(0,s,path,result);
        return result;
    }
    void func(int index,string s,vector<string>& path,vector<vector<string>>& result){
        if(index==s.size()){
            result.push_back(path);
            return;
        }
        for(int i=index;i<s.size();i++){
            if(isPalindrome(s,index,i)){
                path.push_back(s.substr(index,i-index+1));
                func(i+1,s,path,result);
                path.pop_back();
            }
        }
    }
    bool isPalindrome(string s,int start,int end){
        while(start<=end){
            if(s[start++]!=s[end--]) return false;
        }
        return true;
    }
};

File: 131-palindrome-partitioning/131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<>();
        List<String> curr = new ArrayList<>();
        backtrack(result, curr, 0, s);
        return result;
    }
    public void backtrack(List<List<String>> result, List<String> curr, int index, String s){
        if(index == s.length()){
            result.add(new ArrayList<>(curr));
            return;
        }
        for(int i=index; i<s.length(); i++){
            if(isPalindrome(index, i, s)){
                String substring = s.substring(index, i+1);
                curr.add(substring);
                backtrack(result, curr, i+1, s);
                curr.remove(curr.size()-1);
            }
        }
    }
    public boolean isPalindrome(int start, int end, String s){
        while(start<=end){
            if(s.charAt(start++) != s.charAt(end--)) return false;
        }
        return true;
    }
}